Mage.Sets/src/mage/cards/k/KharnTheBetrayer.java

```diff
@@ -0,0 +1,142 @@
+
+        // If damage would be dealt to {this}, prevent that damage and an opponent of your choice gains control of it.
+        Ability replacementAbility = new SimpleStaticAbility(new KharnTheBetrayerEffect());
+        this.addAbility(replacementAbility.withFlavorWord("The Betrayer"));
+    }
+
+    protected KharnTheBetrayer(final KharnTheBetrayer card) {
+        super(card);
+    }
+
+    @Override
+    public Card copy() {
+        return new KharnTheBetrayer(this);
+    }
+    
+}
+
+class KharnTheBetrayerEffect extends ReplacementEffectImpl {
+
+    public KharnTheBetrayerEffect() {
+        super(Duration.WhileOnBattlefield, Outcome.PreventDamage);
+        this.staticText = "If damage would be dealt to {this}, prevent that damage and an opponent of your choice gains control of it.";
+    }
+
+    protected KharnTheBetrayerEffect(final KharnTheBetrayerEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public boolean replaceEvent(GameEvent event, Ability source, Game game) {
+        int damage = game.preventDamage(event, source, game, Integer.MAX_VALUE).getPreventedDamage();
+        Player player = game.getPlayer(source.getControllerId());
+
+        if (player != null && damage > 0) {
+            TargetPlayer target = new TargetOpponent();
+            target.setNotTarget(true);
+            if (!player.choose(outcome, target, source, game)) {
+                return false;
+            }
+            ContinuousEffect effect = new GainControlTargetEffect(Duration.Custom, true, target.getFirstTarget());
+            effect.setTargetPointer(new FixedTarget(source.getSourceId(), game));
+            
+            ReflexiveTriggeredAbility ability 
+                = new ReflexiveTriggeredAbility(effect, false, "an opponent of your choice gains control of it.");
+            game.fireReflexiveTriggeredAbility(ability, source);
+        }
+        return false;
+    }
+
+    @Override
+    public boolean checksEventType(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.DAMAGE_PERMANENT;
+    }
+
+    @Override
+    public boolean applies(GameEvent event, Ability source, Game game) {
+        return event.getTargetId().equals(source.getSourceId());
+    }
+
+    @Override
+    public ContinuousEffect copy() {
+        return new KharnTheBetrayerEffect(this);
+    }
+    
+}
+
+class KharnTheBetrayerTriggeredAbility extends TriggeredAbilityImpl {
+
+    KharnTheBetrayerTriggeredAbility () {
+        super(Zone.BATTLEFIELD, new DrawCardSourceControllerEffect(2));
+    }
+
+    KharnTheBetrayerTriggeredAbility(KharnTheBetrayerTriggeredAbility ability) {
+        super(ability);
+    }
+    
+    @Override
+    public boolean checkEventType(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.LOST_CONTROL;
+    }
+
+    @Override
+    public boolean checkTrigger(GameEvent event, Game game) {
+        return event.getPlayerId().equals(controllerId);
```

Comment:
If my read on the situation is right, when the trigger happens, the player that should draw the cards is not the current controller of Kharn, since the change of control was already done for the event to fire.

So I would advise something like:
-> do not check the player, event.getPlayerId() is the player that should have a trigger to draw 2 cards.
-> there should be a check that the event is for the Kharn source of the Trigger, or else there would be triggers for control change on others permanents.
-> I am really not sure on that, but I think you can change the `controllerId` through `setControllerId(..)` right before returning true? I'd really like a more experimented dev opinion on that. Or maybe some DelayedTriggeredAbility needs to happen somehow.